Wiki-enu-flu-conversion (#4009)
cyberbotics/webots: docs/wiki_flu.md

# New axis system for geometries, devices, and PROTOs

Starting from the R2022a version, all geometries, devices, and PROTOs that come with Webots use the FLU axis orientation (x-**F**orward, y-**L**eft, and z-**U**p, see [#3073](https://github.com/cyberbotics/webots/pull/3073)).
We introduced the update to be consistent with ROS (see [REP 103](https://www.ros.org/reps/rep-0103.html)) and other robotics systems.
If your Webots world uses Webots geometries, devices, and PROTOs then you will probably need to adapt your world.

> [add image example of a 2021b world with the axis and a 2022a world with the axis]

Reviewer: It would be nice to actually add this image now.

Author: ![spot2](https://user-images.githubusercontent.com/48200998/145968280-5579900f-949d-49e5-8e69-d9c0f0f48b32.png)
Example of spot.wbt in 2021b: spot is RUB and the world is NUE.
![spot1](https://user-images.githubusercontent.com/48200998/145968288-5f697a88-68c0-4267-b569-582605702acb.png)
Example of spot.wbt in 2022a: spot is FLU and the world is ENU.

# How to adapt your world or PROTO to Webots R2022a

## Backward compatibility mechanism

> We introduced a backward compatibility mechanism (see [#3619](https://github.com/cyberbotics/webots/pull/3619)) to adapt worlds automatically. The mechanism will be launch when you open a world in the R2022a version.

Reviewer: ```suggestion
We introduced a backward compatibility mechanism (see [#3619](https://github.com/cyberbotics/webots/pull/3619)) to adapt worlds automatically.
This mechanism will be launched when you open a world in the R2022a version.
```

| [Cone](reference/cone.md) | (-$\frac{\pi}{2}$, 0, 0) | B1 |
| [Plane](reference/plane.md) | (-$\frac{\pi}{2}$, 0, 0) | B1 |
| [Camera](reference/camera.md) | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$) | A2 |
| [Lidar](reference/lidar.md) | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$) | A2 |
| [Radar](reference/radar.md) | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$) | A2 |
| [Viewpoint](reference/viewpoint.md) | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$) | A1 |
| [Track](reference/track.md) | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$) | A2 |
| [Pen](reference/camera.md) | (-$\frac{\pi}{2}$, 0, 0) | A2 |
| [Emitter](reference/emitter.md) | (-$\frac{\pi}{2}$, 0, -$\frac{\pi}{2}$) | A2 |
| [Receiver](reference/receiver.md) | (-$\frac{\pi}{2}$, 0, -$\frac{\pi}{2}$) | A2 |
| [Connector](reference/connector.md) | (-$\frac{\pi}{2}$, 0, -$\frac{\pi}{2}$) | A2 |
| [TouchSensor](reference/touchsensor.md) | (-$\frac{\pi}{2}$, 0, -$\frac{\pi}{2}$) | A2 |
| Webots PROTOs | (-$\frac{\pi}{2}$, 0, $\frac{\pi}{2}$), but there are numerous exceptions to this rule as we haven't followed any specific convention | C |

## Automatic conversion to ENU/FLU

> This script `scripts/converter/convert_nue_to_enu_rub_to_flu.py` indents to help you to adapt your world or PROTOs from an old version to the new version of Webots.

Reviewer: We should add a hyperlink to the script in the develop branch.
```suggestion
This script `scripts/converter/convert_nue_to_enu_rub_to_flu.py` intends to help you to adapt your world or PROTOs from an old version to the new version of Webots.
```

> You can find detailed explication into the script (dependencies, usage, limitations and conversion process).

Reviewer: This was not addressed.

[add image example of a "broken" RUB (2021b) to FLU (R2022a) complex proto]

> [add image example of a "broken" NUE (2021b) to ENU (R2022a) complex proto]

Author: ![vil1](https://user-images.githubusercontent.com/48200998/145968884-a52c67b0-056c-480b-9a62-d40057b9a01f.png)
Example of a complex world after backward compatibility algorithm.
![vil2](https://user-images.githubusercontent.com/48200998/145969111-507bfca1-0470-4d03-91e4-5e227645ddb9.png)
Example of a complex world after the execution of the script.

> [add image example of a "broken" NUE (2021b) to ENU (R2022a) complex proto]

Author: other proposition:
![highway](https://user-images.githubusercontent.com/48200998/145969592-30e765f0-0aae-4ccc-b2c5-ba91b23f2fdc.png)
![highway2](https://user-images.githubusercontent.com/48200998/145969600-868c4bda-98aa-4e8e-bc0e-e3556d70facc.png)

![city](https://user-images.githubusercontent.com/48200998/145969722-e059aca8-06d2-43f7-8387-79d09bafc1d3.png)
156.png)
![city2](https://user-images.githubusercontent.com/48200998/145969736-765dbf05-78c8-46c7-9f68-39ce4d78879e.png)
c2fd

![hall](https://user-images.githubusercontent.com/48200998/145969611-a5b87fa2-ca1c-478e-9256-52273549645d.png)
![hall2](https://user-images.githubusercontent.com/48200998/145969699-51c0a43a-07ba-4cdd-b2f6-be209)

![complete_appartment1](https://user-images.githubusercontent.com/48200998/145969750-58d89e9a-1179-46e3-aa6a-878a94715425.png)
![complete_appartment2](https://user-images.githubusercontent.com/48200998/145969760-44988c57-c7c5-4626-89e7-75f28dd3faed.png)

> [add image example of a "broken" NUE (2021b) to ENU (R2022a) complex proto]

Reviewer: The first one looks good to me.

> This script will convert your world from _NUE_ to _ENU_ and the objects contained in the world from _RUB_ to _FLU_. It could be needed to rotate some parts by hand. 

Reviewer: This was not addressed.

Simple usage:
```
python3 convert_nue_to_enu_rub_to_flu.py /your_path_to_your_projects/worlds/my_world.wbt
```

> If your PROTO is _RUB_ (x-**R**ight, y-**U**p, z-**B**ack) and does not contain JavaScript or Lua code, you can convert it to _FLU_ using the same script. You may have to rotate some parts of your PROTOs by hand.

Reviewer: ```suggestion
If your PROTO is _RUB_ (x-**R**ight, y-**U**p, z-**B**ack) and does not contain any JavaScript or Lua code, you can convert it to _FLU_ using the same script. You may have to rotate some parts of your PROTOs manually.
```

**Notes:** check that the conversion went successfully by displaying the different rendering `View/Optional rendering`. In addition, we advise you to check the differences of your `.wbt` or `.proto` before and after the script in a text comparator.
# New axis system recommendations

> As of Webots R2022a, we recommend the FLU (x-**F**orward, y-**L**eft, and z-**U**p) axis orientation for objects and ENU (x-**E**ast, y-**N**orth, and z-**U**p, see [Axes conventions
> ](https://en.wikipedia.org/wiki/Axes_conventions)) for worlds.

Reviewer: ```suggestion
As of Webots R2022a, we recommend the FLU (x-**F**orward, y-**L**eft, and z-**U**p) axis orientation for objects and ENU (x-**E**ast, y-**N**orth, and z-**U**p, see [Axes conventions](https://en.wikipedia.org/wiki/Axes_conventions)) for worlds.
```
